aws-lc-rs/src/key_wrap.rs

```diff
@@ -0,0 +1,424 @@
+
+    /// Returns the algorithm key length.
+    #[inline]
+    #[must_use]
+    fn key_len(&self) -> usize {
+        self.key_len
+    }
+}
+
+impl Sealed for AesBlockCipher {}
+
+impl Debug for AesBlockCipher {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self.id, f)
+    }
+}
+
+/// AES Block Cipher with 128-bit key.
+pub const AES_128: AesBlockCipher = AesBlockCipher {
+    id: BlockCipherId::Aes128,
+    key_len: 16,
+};
+
+/// AES Block Cipher with 256-bit key.
+pub const AES_256: AesBlockCipher = AesBlockCipher {
+    id: BlockCipherId::Aes256,
+    key_len: 32,
+};
+
+/// A Key Wrap (KW) algorithm implementation.
+#[allow(clippy::module_name_repetitions)]
+pub trait KeyWrap: Sealed {
+    /// Peforms the key wrap encryption algorithm using a block cipher.
+    /// It wraps `plaintext` and writes the corresponding ciphertext to `output`.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Any error that has occurred performing the operation.
+    fn wrap<'output>(
+        self,
+        plaintext: &[u8],
+        output: &'output mut [u8],
+    ) -> Result<&'output mut [u8], Unspecified>;
+
+    /// Peforms the key wrap decryption algorithm using a block cipher.
+    /// It unwraps `ciphertext` and writes the corresponding plaintext to `output`.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Any error that has occurred performing the operation.
+    fn unwrap<'output>(
+        self,
+        ciphertext: &[u8],
+        output: &'output mut [u8],
+    ) -> Result<&'output mut [u8], Unspecified>;
+}
+
+/// A Key Wrap with Padding (KWP) algorithm implementation.
+#[allow(clippy::module_name_repetitions)]
+pub trait KeyWrapPadded: Sealed {
+    /// Peforms the key wrap padding encryption algorithm using a block cipher.
+    /// It wraps and pads `plaintext` writes the corresponding ciphertext to `output`.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Any error that has occurred performing the operation.
+    fn wrap_with_padding<'output>(
+        self,
+        plaintext: &[u8],
+        output: &'output mut [u8],
+    ) -> Result<&'output mut [u8], Unspecified>;
+
+    /// Peforms the key wrap padding decryption algorithm using a block cipher.
+    /// It unwraps the padded `ciphertext` and writes the corresponding plaintext to `output`.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Any error that has occurred performing the operation.
+    fn unwrap_with_padding<'output>(
+        self,
+        ciphertext: &[u8],
+        output: &'output mut [u8],
+    ) -> Result<&'output mut [u8], Unspecified>;
+}
+
+/// NIST SP 800-38F key-wrap algorithms.
+///
+/// The NIST specification is similar to that of RFC 3394 but with the following caveats:
+/// * Specifies a maxiumum plaintext length that can be accepted.
+/// * Allows implementations to specify a subset of valid lengths accepted.
+/// * Allows for the usage of other 128-bit block ciphers other than AES.
```

Comment:
The `KeyEncryptionKey` type has a generic parameter `Cipher: BlockCipher` which indicates that the parameter must be any type that implements the `BlockCipher` trait. `AesBlockerCipher` implements `BlockCipher` and other required super traits, making it a valid concrete type that can be used. Then we are implementing the KeyWrap and KeyWrapped padded traits on `KeyEncryptionKey` via `impl KeyWrap for KeyEncryptionKey<AesBlockerCipher>` so that we have the concrete implementations of those traits methods when `KeyEncryptionKey` is used with an `AesBlockCipher`.

What that provides us is that if/when we extend support for other block ciphers or key wrap specifications (like X9.102), we will have a system such that each specification would have a concrete Kek, that Kek can implement the specification for whatever block ciphers it wants to support, it follows the same trait interfaces (`KeyWrap` and/or `KeyWrapPadded` as other specifications), and it is very explicit in the type system and documentation of what specific block ciphers are supported and can be used within a given specification.

Reply:
ah, i see. so this leaves the door open for adding an e.g. `impl KeyWrap for KeyEncryptionKey<ChaCha20BlockerCipher> { ... }` if we need to in the future.

in the case of implementing a new  key wrap spec, is the idea that we add a new struct on which to implement the relevant `KeyWrap` and `KeyWrapPadded` traits? something like

```
    pub struct KeyEncryptionKeyX9_102<Cipher: BlockCipher> {
        cipher: &'static Cipher,
        key: Box<[u8]>,
        other_field_required_by_spec: SomeType,
    }
```

Reply:
Your understanding is correct.